cmd/runtime_options.go

```diff
@@ -52,13 +53,21 @@ func runtimeOptionFlagSet(includeSysEnv bool) *pflag.FlagSet {
 	flags := pflag.NewFlagSet("", 0)
 	flags.SortFlags = false
 	flags.Bool("include-system-env-vars", includeSysEnv, "pass the real system environment variables to the runtime")
+	// NOTE(imiric): This flag should only be defined in cmd.optionFlagSet,
+	// yet because of the fragmented JS runtime options configuration
+	// (and flag handling?) the cmd.TestEnvVars test would fail if this
+	// wasn't redefined here. Functionally it shouldn't matter, but this
+	// should be fixed as part of #883.
+	flags.String("compatibility-mode", "es6",
+		`JavaScript compiler compatibility mode, "es6" or "es51"`)
```

Comment:
I don't think the default value should be called `es6` - after all, currently k6 supports more than ECMAScript 6 features. And in the future, we'll update goja/core.js/babel to support even newer ECMAScript versions, so we'd definitely have to change the default value to be something different than `es6`, which isn't ideal. Instead, the default value should be something like `latest` or even simply `default`, meaning that we run the script with the latest ES version k6 supports.

`es51` also seems like it should be `es5.1` - given that we're on `es10` now, the decimal point seems necessary :sweat_smile:

Reply:
I also have a problem with `es5.1` or `es51`. In the future we might decide to enhance goja to support more than just es51(which it supposedly supports more than ATM) ... maybe es6 maybe more, at which point it won't be the correct name as well. 
I am for the much less "standard names", maybe something like `extended` and `base` will be a lot better with explanations on what those are.

Reply:
> after all, currently k6 supports more than ECMAScript 6 features

You mean because of core.js also having shims for ES7? That's true.

OK, so how about we adopt [the same model as core.js](https://stackoverflow.com/a/45568325/96213) then: split the modes apart and let the user choose. We make the flag `--js-compatibility` be a list of modes: "es51", "es6", "es7", "esfinal10", etc. So the default right now would be `{"es51", "es6", "es7"}`. We would have to split apart the current core.js, but that shouldn't be a problem. I'm unsure what the behavior should be when the list is empty :smile:

> the default value should be something like `latest` or even simply `default`

That would mean compatibility is always a moving target and the user wouldn't know what to expect. While the default mode list would still change occasionally, we should still warn in advance when this happens and encourage in docs to set the list of modes explicitly and not rely on the default behavior.

Reply:
> That would mean compatibility is always a moving target and the user wouldn't know what to expect. 

Well... two things:
1. Newer ES versions are generally backwards compatible, so user scripts should continue to work without issues
2. Generally, newer browsers and node.js versions do exactly this - they just silently start supporting newer ES features :wink: So it's not like k6 would do something completely unheard of... Our only difference is that we want to add a mode that turns off that support, for some performance gains. 

So, in my mind, it makes perfect sense to currently have 2 possible compatibility mode values:
- `latest` or `default` - enabled babel and core.js, basically support everything ES that k6 can support
- `es5.1` - support only what goja currently supports (and even that might be problematic, if we decide to invest time extending goja, but we'll burn that bridge when we come to it...)

Reply:
To be clear, the goal of this feature isn't to be able to have a very fine-tuned core.js/babel configuration, and to support very distinct ES versions for it's own sake. Or at least I don't see a very valid use case for that... The current goal is to have a high-performance mode for k6 where each VU doesn't take several MB of RAM because of core.js...

We might extend it in the future to support adjacent things like TypeScript (via the Babel plugin), but for now at least I don't see the point of supporting every single ES version individually. And if it turns out that there's a valid use case for that, we can always add support for it in the future... though passing raw JSON in CLI flags probably won't be the way, rather we can probably configure a path to a `.babelrc` file or something like that... In any way, that'd be far in the future, if ever.

Reply:
(I think you meant _cross_ that bridge? We probably shouldn't burn it ;))

@MStoykov has issues with `es5.1` (and even Goja advertises as being "5.1(+)"), but we shouldn't bikeshed too much about this. Like you say, we'll figure out the Goja extensions mode name when we get to it, though that should mostly be following the ES specs and should be transparent to the user, just be more performant when we drop Babel or core.js.

So, decided: I'll change it to `latest` (default) and `es5.1`.

Reply:
This should be resolved by bae2ae26.

The only problem is that I didn't find a way to use "es5.1" with enumer, so I truncated it at "es5", which gets the point across just as well.

Let me know.